gap/hashfun.gi

```diff
@@ -0,0 +1,383 @@
+    # "official" length, therefore we *must not* use the last, half-used
+    # byte. This inevitably leads to collisions!
+
+# TODO: verify this. possibly work around it in GAP...
+    if bytelen <= 8 then
+        return rec( func := DATA_HashFunctionForShort8BitVectors,
+                    data := [hashlen,q] );
+    else
+        return rec( func := DATA_HashFunctionFor8BitVectors,
+                    data := [hashlen,bytelen] );
+    fi;
+end );
+
+#
+# uncompressed finite field vectors
+#
+InstallMethod( ChooseHashFunction, "for finite field vectors over big finite fields",
+    [IsList],
+function(l)
+    local f, q;
+    if NestingDepthA(l) = 1 and Length(l) > 0 and IsFFE(l[1]) then
+        # FIXME: so this assumes the list is homogeneous, and then
+        # blindly asks for a list? Perhaps we should at least
+        # check for IsHomogeneousList and/or add that to the filters list?
+        f := Field(l);
+        q := Size(f);
+        return rec( func := DATA_HashFunctionForShort8BitVectors,
+                    data := [hashlen,q] );
+    fi;
+    TryNextMethod();
+end );
+
+
+#
+# compressed matrices
+#
+InstallGlobalFunction( DATA_HashFunctionForCompressedMats,
+function(mat)
+    local i, res;
+
+# FIXME: do not do this. Instead, properly chain
+# HashKeyBag invocations, by using the seed argument:
+if false then
+    # HACKISH prototype; does not take short vectors into account,
+    # nor GF(2) vs other fields, etc.
+    res := 101;
+    offset := 3 * GAPInfo.BytesPerVariable;
+TODO
+    bytelen := 100;
+    for i in [1 .. Length(mat)] do
+        HashKeyBag(mat[i], res, offset, bytelen);
+    od;
+
+    end
+fi;
+
+# Alternatively, compute the separate hashes for each row,
+# then compute hash of the resulting vector (but this
+# introduces more temporaries, so is not as nice)
+
+  res := 0;
+  for i in [1..Length(mat)] do
+      res := (res * data[3] + data[2].func(mat[i],data[2].data)) mod data[1];
+  od;
+  return res + 1;
+end );
+
+InstallMethod( ChooseHashFunction, "for compressed gf2 matrices",
+    [IsGF2MatrixRep and IsList],
+function(p)
+    local data;
+    data := [hashlen,ChooseHashFunction(p[1],hashlen),
+             PowerMod(2,Length(p[1]),hashlen)];
+    return rec( func := DATA_HashFunctionForCompressedMats,
+                data := data );
+  end );
+
+InstallMethod( ChooseHashFunction, "for compressed 8bit matrices",
+    [Is8BitMatrixRep and IsList],
+function(p)
+    local data,q;
+    q := Q_VEC8BIT(p[1]);
+    data := [hashlen,ChooseHashFunction(p[1],hashlen),
+             PowerMod(q,Length(p[1]),hashlen)];
+    return rec( func := DATA_HashFunctionForCompressedMats,
+                data := data );
+  end );
+
+
+#
+# Integers
+#
+InstallGlobalFunction( DATA_HashFunctionForIntegers,
+function(x)
+    # TODO: implement this on the C level, to return
+    # the integer itself it is a small integer; or else
+    # to return the integer mod 2^N so that it fits into
+    # a small integer
```

Comment:
Note that this is just a fast default. Of course there may be applications where you would want to hash integers differently. E.g. if they are all multiples of 2^64, then this approach would give the hash value 0 for all of them.
So an alternative would be to run `HashKeyBag` on them.